test/lit/passes/gsi.wast

```diff
@@ -1268,3 +1268,162 @@
     )
   )
 )
+
+;; Two subtypes, each with a global. A get of the parent can be optimized into
+;; a select, as it must read one of the children.
```

Comment:
This should be gated behind --closed-world, but it doesn't look like it is. With an open world, some other value with the same type may have flowed in.

Reply:
Still paging back here, but wouldn't that be an issue only for a public type in closed world mode?

Reply:
No, because in closed world mode the only public types are function types (at least for now). If we had a public struct type in closed world mode, then I think we would still want the closed world assumption that only the current module can produce values of that type and that the environment could still only hold on to them and pass them back in but not inspect or produce them. The exception would be for the public types of imported globals, which must be produced by the environment. So really there are three levels of visibility in a generalized closed world mode:

1. imported public - There is an import with this type, so the environment must be able to create and inspect values with this type. This is only possible for function types under the current closed-world validation rules.
2. public - The type identity must be preserved so that the module signature does not change, but we can still assume that the environment does not produce or inspect values of this type. This visibility is not allowed under the current closed-world validation rules.
3. private - Arbitrary type modifications are allowed.

This pass might need to updated to handle "imported public" struct types correctly in the future. I'm not sure.

Reply:
I see. Yeah, more to think about here, long term.

For now this pass only runs in closed world mode, I see, so I think we are ok.

https://github.com/WebAssembly/binaryen/blob/e761198190d5c5e6e0a3735c3dab59d02c970e43/src/passes/pass.cpp#L625-L628

Reply:
Ah, then it looks like I missed adding an error to this pass to validate that we're running in closed world mode in #5367. Would you mind adding that error here and updating the test?

Reply:
Is there a reason to do it in this PR? I can do it separately.

Reply:
Oh, I guess it will conflict immediately in 2 places. I'll do it here...

Reply:
Unfortunately it's not simple, as several testcases need nontrivial modifications. I'll leave it out of this PR and look into it separately.

Reply:
Sounds good 👍